src/macaron/repo_finder/repo_finder.py

```diff
@@ -0,0 +1,182 @@
+    repo-based type in https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst.
+
+    Note that this method will be updated when there are new pre-defined types as per the PURL specification.
+
+    Parameters
+    ----------
+    purl_type : str
+        The type field of the PURL.
+
+    Returns
+    -------
+    str | None
+        The git service domain corresponding to the purl type or None if the purl type is unknown.
+    """
+    known_types = {"github": "github.com", "bitbucket": "bitbucket.org"}
+    return known_types.get(purl_type, None)
+
+
+def to_repo_path(purl: PackageURL, available_domains: list[str]) -> str | None:
+    """Return the repository path from the PURL string.
+
+    This method only supports converting a PURL with the following format:
+
+        pkg:<type>/<namespace>/<name>[...]
+
+    Where ``type`` could be either:
+    - The pre-defined repository-based PURL type as defined in
+    https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst
+
+    - The supprted git service domains (e.g. ``github.com``) defined in ``available_domains``.
+
+    The repository path will be generated with the following format ``https://<type>/<namespace>/<name>``.
+
+    Parameters
+    ----------
+    purl : PackageURL
+        The parsed PURL to convert to the repository path.
+    available_domains: list[str]
+        The list of available domains
+
+    Returns
+    -------
+    str | None
+        The URL to the repository which the PURL is referring to or None if we cannot convert it.
+    """
+    domain = to_domain_from_known_purl_types(purl.type) or (purl.type if purl.type in available_domains else None)
+    if not domain:
+        logger.error("The PURL type of %s is not valid as a repository type.", purl.to_string())
+        # Try to find the repository
+        return find_repo(purl)
+
+    if not purl.namespace:
+        logger.error("Expecting a non-empty namespace from %s.", purl.to_string())
+        return None
+
+    # TODO: Handle the version tag and commit digest if they are given in the PURL.
+    return urlunparse(
+        ParseResult(
+            scheme="https",
+            netloc=domain,
+            path=os.path.join(purl.namespace, purl.name),
+            params="",
+            query="",
+            fragment="",
+        )
+    )
+
+
+def find_valid_url(urls: Iterable[str]) -> str:
+    """Find a valid URL from the provided URLs.
+
+    Parameters
+    ----------
+    urls : Iterable[str]
+        An Iterable object containing urls.
+
+    Returns
+    -------
+    str
+        A valid URL or empty if it can't find any valid URL.
+    """
+    vcs_set = {get_remote_vcs_url(value) for value in urls if get_remote_vcs_url(value) != ""}
+
+    # To avoid non-deterministic results we sort the URLs.
+    vcs_list = sorted(vcs_set)
+
+    if len(vcs_list) < 1:
+        return ""
+
+    # Report the first valid URL.
+    return vcs_list.pop()
```

Comment:
`list.pop()` returns the last element in `list`, not the first one ([Python doc](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists)).
I suppose this is probably what you want?
```suggestion
    return vcs_list[0]
```

Reply:
This function was moved from `dependency_resolver.py` and has remained unchanged.

Reply:
Oh, I see. Up to you to decide whether to change it. 👍🏼

Reply:
@behnazh-w @tromai Was the original intent of this to return the first element as the comment suggests, or the last as the code performs?

Reply:
The issue is in my comment, which should be the last element not first.